File: Assignment 3/dilsin_amirtharajah_assignment3.cpp
#include <iostream> 
#include <vector>   
#include <limits>   // for input
using namespace std;

typedef vector<vector<int>> Matrix; // matrix type

Matrix input_matrix(int rows, int cols) { // input matrix
    Matrix mat; // empty list for matrix
    for (int i = 0; i < rows; i++) { // loop through rows
        while (true) {
            cout << "Enter values for Matrix (row " << i + 1 << " of " << rows 
                 << ", " << cols << " integers separated by spaces): ";
            Matrix temp(1); 
            temp[0].resize(cols); 
            bool valid = true; 

            for (int j = 0; j < cols; j++) {
                if (!(cin >> temp[0][j])) { // if input is not integer
                    cout << "Invalid input! Please enter a number." << endl; 
                    cin.clear(); 
                    cin.ignore(numeric_limits<streamsize>::max(), '\n'); 
                    valid = false; 
                    break; 
                }
            }

            if (valid) { 
                mat.push_back(temp[0]); // add row to matrix
                break; 
            }
        }
    }
    return mat; 
}

void print_matrix(const Matrix &mat) { // print matrix
    for (auto &row : mat) { // loop through each row
        for (auto &val : row) cout << val << " "; 
        cout << endl;
    }
}

Matrix add_matrix(const Matrix &A, const Matrix &B) { // add matrix
    Matrix result(A.size(), vector<int>(A[0].size()));
    for (int i = 0; i < (int)A.size(); i++) // loop rows i
        for (int j = 0; j < (int)A[0].size(); j++) // loop columns j
            result[i][j] = A[i][j] + B[i][j]; // add
    return result;
}

Matrix subtract_matrix(const Matrix &A, const Matrix &B) { // subtract matrix
    Matrix result(A.size(), vector<int>(A[0].size()));
    for (int i = 0; i < (int)A.size(); i++) // loop rows i
        for (int j = 0; j < (int)A[0].size(); j++) // loop columns j
            result[i][j] = A[i][j] - B[i][j]; // subtract
    return result;
}

Matrix transpose_matrix(const Matrix &A) { // transpose matrix
    Matrix result(A[0].size(), vector<int>(A.size()));
    for (int i = 0; i < (int)A[0].size(); i++) // loop columns i
        for (int j = 0; j < (int)A.size(); j++) // loop rows j
            result[i][j] = A[j][i]; // swap
    return result;
}

Matrix multiply(const Matrix &A, int scalar = 2) { // multiply by scalar
    Matrix result(A.size(), vector<int>(A[0].size()));
    for (int i = 0; i < (int)A.size(); i++) // loop rows i
        for (int j = 0; j < (int)A[0].size(); j++) // loop columns j
            result[i][j] = A[i][j] * scalar; // multiply
    return result;
}

int get_input(string prompt, int low = -1, int high = -1) { // input
    int val;
    while (true) {
        cout << prompt;
        if (!(cin >> val)) { // try convert input
            cout << "Invalid input! Please enter a number." << endl; 
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
        } else if (low != -1 && high != -1 && (val < low || val > high)) {
            cout << "Invalid input! Please enter a number between " << low << " and " << high << "." << endl; // error if out of range
        } else {
            return val; // return valid integer
        }
    }
}

void main_program() { // main 
    while (true) { // keep running until quit
        cout << "\nWelcome to the Matrix Calculator!" << endl;

        int rows = get_input("Enter number of rows (1-5): ", 1, 5); // read rows
        int cols = get_input("Enter number of columns (1-5): ", 1, 5); // read columns

        cout << "\nEnter values for Matrix A:" << endl; // ask for matrix A
        Matrix A = input_matrix(rows, cols); // input A

        cout << "\nEnter values for Matrix B:" << endl; // ask for matrix B
        Matrix B = input_matrix(rows, cols); // input B

        cout << "\nChoose an operation:" << endl; // menu text
        cout << "1. Add" << endl; // option 1
        cout << "2. Subtract" << endl; // option 2
        cout << "3. Transpose A" << endl; // option 3
        cout << "4. Scalar Multiply A" << endl; // option 4

        int choice;
        while (true) { // check menu choice
            choice = get_input("Your choice: ");
            if (choice < 1 || choice > 4) // check valid
                cout << "Invalid input! Please enter a number between 1 and 4." << endl;
            else
                break;
        }

        cout << endl;
        if (choice == 1) { // add
            cout << "Result of A + B:" << endl;
            print_matrix(add_matrix(A, B));
        } else if (choice == 2) { // subtract
            cout << "Result of A - B:" << endl;
            print_matrix(subtract_matrix(A, B));
        } else if (choice == 3) { // transpose
            cout << "Result of A^T:" << endl;
            print_matrix(transpose_matrix(A));
        } else if (choice == 4) { // multiply
            int s = get_input("Enter scalar value: ");
            cout << "Result of " << s << " * A:" << endl; // result text
            print_matrix(multiply(A, s)); // print result
        }

        cout << endl;
        while (true) {
            string again;
            cout << "Would you like to try again? (y/n): ";
            cin >> again;
            if (again == "y") {
                break; 
            } else if (again == "n") {
                cout << "Thanks for using the Matrix Calculator. Goodbye!" << endl;
                return; // exit
            } else {
                cout << "Invalid input! Please enter 'y' for yes or 'n' for no." << endl; // error
            }
        }
    }
}

int main() {
    main_program();
    return 0;
}

File: Assignment 3/dilsin_amirtharajah_assignment3.py
def input_matrix(rows, cols): # input matrix
    mat = [] # empty list for matrix
    for i in range(rows): # loop through rows
        while True: 
            try: # convert
                row = list(map(int, input(f"Enter values for Matrix (row {i+1} of {rows}, {cols} integers separated by spaces): ").split())) # input reads row values, split splits into strings, map convert to int, list stores as list
                if len(row) != cols: # check if correct columns
                    print(f"Invalid input! Please enter exactly {cols} integers.") # error if wrong 
                else:
                    mat.append(row) # add row to matrix
                    break # exit loop 
            except ValueError: # if input is not integer
                print("Invalid input! Please enter a number.") 
    return mat # 

def print_matrix(mat): # print matrix
    for row in mat: # loop through each row
        print(*row) # print row with spaces

def add_matrix(A, B): # add matrix
    return [[A[i][j] + B[i][j] for j in range(len(A[0]))] for i in range(len(A))] # loop rows i, loop columns j, add

def subtract_matrix(A, B): # subtract matrix
    return [[A[i][j] - B[i][j] for j in range(len(A[0]))] for i in range(len(A))] # loop rows i, loop columns j, subtract

def transpose_matrix(A): # transpose matrix
    return [[A[j][i] for j in range(len(A))] for i in range(len(A[0]))] # loop columns i, loop rows j, swap

def multiply(A, scalar=2): # multiply by scalar
    return [[A[i][j] * scalar for j in range(len(A[0]))] for i in range(len(A))] # loop rows i, loop columns j, multiply

def get_input(prompt, low=None, high=None): # input
    while True: 
        try: # try convert input
            val = int(input(prompt)) # read integer
            if low is not None and high is not None: 
                if val < low or val > high: # check range
                    print(f"Invalid input! Please enter a number between {low} and {high}.") # error if out of range
                    continue # ask 
            return val # return valid integer
        except ValueError: # if not an integer
            print("Invalid input! Please enter a number.") # error message

def main(): # main 
    while True: # keep running until quit
        print("\nWelcome to the Matrix Calculator!") 

        rows = get_input("Enter number of rows (1–5): ", 1, 5) # read rows
        cols = get_input("Enter number of columns (1–5): ", 1, 5) # read columns

        print() 
        print("Enter values for Matrix A:") # ask for matrix A
        A = input_matrix(rows, cols) # input A

        print() 
        print("Enter values for Matrix B:") # ask for matrix B
        B = input_matrix(rows, cols) # input B

        print() 
        print("Choose an operation:") # menu text
        print("1. Add") # option 1
        print("2. Subtract") # option 2
        print("3. Transpose A") # option 3
        print("4. Scalar Multiply A") # option 4

        while True: # check menu choice
            try:
                choice = int(input("Your choice: ")) # read choice
                if choice not in [1, 2, 3, 4]: # check valid
                    print("Invalid input! Please enter a number between 1 and 4.") 
                else:
                    break 
            except ValueError: # not integer
                print("Invalid input! Please enter a number.") # error 

        print() 
        if choice == 1: # add
            print("Result of A + B:") 
            print_matrix(add_matrix(A, B)) 
        elif choice == 2: # subtract
            print("Result of A - B:") 
            print_matrix(subtract_matrix(A, B)) 
        elif choice == 3: # transpose
            print("Result of A^T:") 
            print_matrix(transpose_matrix(A)) 
        elif choice == 4: # multiply
            while True: 
                try:
                    s = int(input("Enter scalar value: ")) # read scalar
                    break 
                except ValueError: # not integer
                    print("Invalid input! Please enter a number.") 
            print(f"Result of {s} * A:") # result text
            print_matrix(multiply(A, s)) # print result

        print() 
        while True: 
            again = input("Would you like to try again? (y/n): ").strip().lower() 
            if again == "y": 
                break # restart 
            elif again == "n": 
                print("Thanks for using the Matrix Calculator. Goodbye!") 
                return # exit
            else: 
                print("Invalid input! Please enter 'y' for yes or 'n' for no.") # error

main() 
